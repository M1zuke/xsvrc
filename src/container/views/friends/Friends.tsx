import React, { ReactElement, useCallback, useEffect, useMemo, useState } from 'react';
import { useSelector } from 'react-redux';
import { ifLoaded } from '../../../api/prepare';
import { useApi } from '../../../api/use-api';
import { Button } from '../../../components/button/Button';
import { Checkbox } from '../../../components/checkbox/Checkbox';
import { FriendOverview } from '../../../components/friend-overview/FriendOverview';
import { Grid } from '../../../components/grid/Grid';
import { LoadableContent } from '../../../components/loadable-content/LoadableContent';
import { ScrollableContent } from '../../../components/scrollable-content/ScrollableContent';
import { friendInfo } from '../../../store/friends/selectors';
import styles from './Friends.module.scss';

export const CharacterFilters = [
  'ALL',
  '#',
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
] as const;
export type CharacterFilter = typeof CharacterFilters[number];

export function Friends(): ReactElement {
  const { friends } = useApi();
  const friendsInfos = useSelector(friendInfo);
  const [filter, setFilter] = useState<CharacterFilter>('ALL');
  const [showPrivate, setShowPrivate] = useState(true);

  const friendInfoWithOrWithoutPrivate = useMemo(() => {
    if (ifLoaded(friendsInfos) && !showPrivate) {
      return friendsInfos.filter((fi) => fi.location !== 'private');
    }
    return friendsInfos;
  }, [friendsInfos, showPrivate]);

  const disableFilterButton = useCallback(
    (key: CharacterFilter) => {
      if (!ifLoaded(friendInfoWithOrWithoutPrivate)) {
        return true;
      }

      if (key === 'ALL') {
        return false;
      }

      if (key === '#') {
        const filteredFriendInfo = friendInfoWithOrWithoutPrivate.filter(
          (o) => !CharacterFilters.includes(o.displayName[0].toUpperCase() as CharacterFilter),
        );
        return filteredFriendInfo.length === 0;
      }

      const filteredFriendInfo = friendInfoWithOrWithoutPrivate.filter(
        (o) => o.displayName[0].toLowerCase() === key.toLowerCase(),
      );
      return filteredFriendInfo.length === 0;
    },
    [friendInfoWithOrWithoutPrivate],
  );

  const filteredFriendInfo = useMemo(() => {
    if (!ifLoaded(friendInfoWithOrWithoutPrivate) || filter === 'ALL') {
      return friendInfoWithOrWithoutPrivate;
    }
    if (filter === '#') {
      return friendInfoWithOrWithoutPrivate.filter(
        (o) => !CharacterFilters.includes(o.displayName[0].toUpperCase() as CharacterFilter),
      );
    }
    return friendInfoWithOrWithoutPrivate.filter((fi) => fi.displayName.startsWith(filter));
  }, [filter, friendInfoWithOrWithoutPrivate]);

  const filterButtons = useMemo(() => {
    return CharacterFilters.map((key) => {
      return (
        <Button
          key={`select-filter-button-${key}`}
          aria-label={`select-filter-button-${key}`}
          onClick={() => setFilter(key)}
          active={filter === key}
          disabled={disableFilterButton(key)}
        >
          {key}
        </Button>
      );
    });
  }, [disableFilterButton, filter]);

  useEffect(() => {
    friends().finally();

    const interval = setInterval(() => {
      friends().finally();
    }, 25000);

    return () => {
      clearInterval(interval);
    };
  }, [friends]);

  return (
    <div className={styles.Component}>
      <div className={styles.FilterButtons}>
        <div className={styles.NormalFilter}>{filterButtons}</div>
        <div className={styles.SpecialFilter}>
          <Checkbox label="Show Private" onClick={setShowPrivate} value={showPrivate} />
        </div>
      </div>
      <ScrollableContent>
        <LoadableContent data={filteredFriendInfo}>
          {(data) => (
            <Grid columns={8}>
              {data.map((friendInfo) => (
                <FriendOverview friendInfo={friendInfo} key={`Friends-Overview-${friendInfo.displayName}`} />
              ))}
            </Grid>
          )}
        </LoadableContent>
      </ScrollableContent>
    </div>
  );
}
