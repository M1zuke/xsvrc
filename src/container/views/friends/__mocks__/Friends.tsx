import React from 'react';

export const Friends = jest.fn(() => <div data-testid="friends" />);
