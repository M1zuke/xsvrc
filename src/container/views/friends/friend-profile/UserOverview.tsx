import React, { ReactElement } from 'react';
import { UserInfo } from '../../../../api/types';
import styles from './UserOverview.module.scss';

type UserBioProps = {
  user: UserInfo;
};

export function UserOverview({ user }: UserBioProps): ReactElement {
  return (
    <div className={styles.Component}>
      <div className={styles.UserBio}>
        <pre>{user.bio}</pre>
      </div>
    </div>
  );
}
