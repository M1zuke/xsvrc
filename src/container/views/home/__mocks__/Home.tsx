import React from 'react';

export const Home = jest.fn(() => <div data-testid="home" />);
