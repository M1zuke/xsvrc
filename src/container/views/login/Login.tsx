import React, { ReactElement, useCallback, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { useApi } from '../../../api/use-api';
import { Button } from '../../../components/button/Button';
import { TextInput } from '../../../components/input/TextInput';
import { Loading } from '../../../components/loading/Loading';
import { isProbablyAuthenticated } from '../../../components/secured/Secured';
import { useMessages } from '../../../i18n';
import { savedCookies } from '../../../store/cookies/selectors';
import { vrcUserInfo } from '../../../store/user/selectors';
import styles from './Login.module.scss';

export function Login(): ReactElement {
  const messages = useMessages();
  const cookies = useSelector(savedCookies);
  const userInfo = useSelector(vrcUserInfo);
  const { login } = useApi();
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const handleUsernameChange = useCallback((value: string) => setUsername(value), []);
  const handlePasswordChange = useCallback((value: string) => setPassword(value), []);

  const handleLogin = useCallback(() => {
    login(username, password).finally();
  }, [login, password, username]);

  useEffect(() => {
    if (isProbablyAuthenticated(cookies)) {
      handleLogin();
    }
  }, [cookies, handleLogin]);

  return (
    <div className={styles.Component}>
      <div className={styles.LoginBox}>
        <div className={styles.LoginImage} />
        <TextInput
          aria-label="enter username"
          value={username}
          onChange={handleUsernameChange}
          placeholder={messages.Views.Login.View.Username}
        />
        <TextInput
          aria-label="enter password"
          value={password}
          onChange={handlePasswordChange}
          type="password"
          placeholder={messages.Views.Login.View.Password}
        />
        {userInfo === 'loading' ? (
          <Loading />
        ) : (
          <Button aria-label="login" onClick={handleLogin}>
            {messages.Views.Login.View.Login}
          </Button>
        )}
      </div>
    </div>
  );
}
