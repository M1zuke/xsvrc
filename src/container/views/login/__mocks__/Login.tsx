import React from 'react';

export const Login = jest.fn(() => <div data-testid="login" />);
