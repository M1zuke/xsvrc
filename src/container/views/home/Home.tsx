import React, { ReactElement, useMemo } from 'react';
import { useSelector } from 'react-redux';
import { isLoaded } from '../../../api/prepare';
import { isOnline } from '../../../common/utils';
import { Content } from '../../../components/content/Content';
import { FriendFavoriteList } from '../../../components/friend-favorite-list/FriendFavoriteList';
import { LoadableContent } from '../../../components/loadable-content/LoadableContent';
import { Loading } from '../../../components/loading/Loading';
import { ScrollableContent } from '../../../components/scrollable-content/ScrollableContent';
import { useMessages } from '../../../i18n';
import { selectFriendInfo } from '../../../store/friends/selectors';
import { selectFriendFavorites, selectNotifications, selectUserInfo } from '../../../store/user/selectors';
import styles from './Home.module.scss';
import { TitleBox } from './TitleBox';

export function Home(): ReactElement {
  const userInfo = useSelector(selectUserInfo);
  const friendInfo = useSelector(selectFriendInfo);
  const favorites = useSelector(selectFriendFavorites);
  const allNotifications = useSelector(selectNotifications);
  const messages = useMessages();

  const onlineFriends = useMemo(
    () => (isLoaded(friendInfo) ? Object.values(friendInfo) : []).filter((fi) => isOnline(fi)).length,
    [friendInfo],
  );

  const notifications = useMemo(() => (isLoaded(allNotifications) ? allNotifications : []), [allNotifications]);
  const friendRequests = useMemo(
    () => notifications.filter((not) => not.type === 'friendRequest').length,
    [notifications],
  );
  const invites = useMemo(() => notifications.filter((not) => not.type === 'invite').length, [notifications]);

  if (!isLoaded(userInfo) || !isLoaded(friendInfo) || !isLoaded(favorites)) {
    return (
      <Content translucent className={styles.FullScreen}>
        <Loading />
      </Content>
    );
  }

  return (
    <ScrollableContent className={styles.PaddingOverwrite} innerClassName={styles.Component} translucent>
      <div className={styles.NumberOverview}>
        <TitleBox title={messages.Views.Dashboard.Friends.Online}>{onlineFriends}</TitleBox>
        <TitleBox title={messages.Views.Dashboard.Friends.FriendRequests}>{friendRequests}</TitleBox>
        <TitleBox title={messages.Views.Dashboard.Friends.Invites}>{invites}</TitleBox>
      </div>
      <LoadableContent data={favorites}>
        {(data) => (
          <div className={styles.FriendFavorites}>
            <FriendFavoriteList favorites={data} />
          </div>
        )}
      </LoadableContent>
    </ScrollableContent>
  );
}
