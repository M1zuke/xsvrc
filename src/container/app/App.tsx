import React, { lazy, Suspense, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { Route, Switch } from 'react-router-dom';
import { isLoaded } from '../../api/prepare';
import { useApi } from '../../api/use-api';
import { routes } from '../../common/routes';
import { useSubscribe } from '../../common/use-subscribe';
import { Header } from '../../components/header/Header';
import { Loading } from '../../components/loading/Loading';
import { Navigation } from '../../components/navigation/Navigation';
import { Secured } from '../../components/secured/Secured';
import { WebSockets } from '../../components/websockets/WebSockets';
import { selectUserInfo } from '../../store/user/selectors';
import styles from './App.module.scss';

const HomeComponent = lazy(() => import('../views/home/Home').then(({ Home }) => ({ default: Home })));
const FriendComponent = lazy(() => import('../views/friends/Friends').then(({ Friends }) => ({ default: Friends })));
const FriendProfile = lazy(() =>
  import('../views/friends/friend-profile/FriendProfile').then(({ FriendProfile }) => ({ default: FriendProfile })),
);
const EventList = lazy(() =>
  import('../views/user-event-list/UserEventList').then(({ UserEventList }) => ({ default: UserEventList })),
);
const Notifications = lazy(() =>
  import('../views/notifications/Notifications').then(({ Notifications }) => ({ default: Notifications })),
);

const App: React.FC = () => {
  const userInfo = useSelector(selectUserInfo);
  const { info } = useApi();
  const { getUser } = useApi();
  useSubscribe(getUser, isLoaded(userInfo) ? userInfo.id : null);

  useEffect(() => {
    info().then();
  }, [info]);

  return (
    <div className={styles.Component}>
      <Header />
      <div className={styles.Application}>
        <Navigation />
        <WebSockets />
        <Secured>
          <div className={styles.PageContent}>
            <Suspense fallback={<Loading />}>
              <Switch>
                <Route {...routes.home.config} component={HomeComponent} />
                <Route {...routes.friends.config} component={FriendComponent} />
                <Route {...routes.friendsProfile.config} component={FriendProfile} />
                <Route {...routes.eventList.config} component={EventList} />
                <Route {...routes.notifications.config} component={Notifications} />
              </Switch>
            </Suspense>
          </div>
        </Secured>
      </div>
    </div>
  );
};

export default App;
