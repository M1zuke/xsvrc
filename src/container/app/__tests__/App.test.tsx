import { waitFor } from '@testing-library/dom';
import pretty from 'pretty';
import React from 'react';
import * as ElectronFetchModule from '../../../common/electron-fetch';
import { ElectronResult } from '../../../common/electron-fetch';
import { routes } from '../../../common/routes';
import * as UserOverviewModule from '../../../components/user-overview/UserOverview';
import { VRCApiInfo } from '../../../store/api-info/state';
import { render, withAuthentication } from '../../../test-utils';
import App from '../App';
import * as FriendsModule from '../../views/friends/Friends';
import * as HomeModule from '../../views/home/Home';
import * as LoginModule from '../../views/login/Login';

const { electronFetch: FetchMock } = ElectronFetchModule as jest.Mocked<typeof ElectronFetchModule>;
const { Login: LoginMock } = LoginModule as jest.Mocked<typeof LoginModule>;
const { Home: HomeMock } = HomeModule as jest.Mocked<typeof HomeModule>;
const { Friends: FriendsMock } = FriendsModule as jest.Mocked<typeof FriendsModule>;
const { UserOverview: UserOverviewMock } = UserOverviewModule as jest.Mocked<typeof UserOverviewModule>;

jest.mock('../../../common/electron-fetch');
jest.mock('../../views/login/Login');
jest.mock('../../views/home/Home');
jest.mock('../../views/friends/Friends');
jest.mock('../../../components/user-overview/UserOverview');

describe('<App /> Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders login Component', async () => {
    FetchMock.mockReturnValueOnce(
      new Promise<ElectronResult<VRCApiInfo>>((resolve) => {
        resolve({
          type: 'entity',
          cookies: [],
          result: { clientApiKey: 'testClientApiKey' },
        });
      }),
    );
    const customResult = render(<App />, {
      location: { pathname: routes.login.path },
      route: { path: routes.login.path },
    });

    await waitFor(() => {
      expect(LoginMock).toHaveBeenCalled();
    });

    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
  });

  it('fetches VRCApiInfo', async () => {
    FetchMock.mockReturnValueOnce(
      new Promise<ElectronResult<VRCApiInfo>>((resolve) => {
        resolve({
          type: 'entity',
          cookies: [],
          result: { clientApiKey: 'testClientApiKey' },
        });
      }),
    );
    const customResult = render(<App />, {
      location: { pathname: routes.login.path },
      route: { path: routes.login.path },
    });
    await waitFor(() => {
      expect(FetchMock).toHaveBeenCalledWith({
        storedCookies: [],
        url: 'https://api.vrchat.cloud/api/1/config',
      });
    });

    expect(customResult.store.getState().apiInfo).toEqual({
      clientApiKey: 'testClientApiKey',
    });
  });

  type mockedComponents = typeof HomeMock | typeof FriendsMock;
  const mappedMockComponentsToRoute: Record<string, { path: string; component: mockedComponents }> = {
    home: { component: HomeMock, path: routes.home.path },
    friends: { component: FriendsMock, path: routes.friends.path },
  };

  it.each(Object.values(mappedMockComponentsToRoute))('renders %s Component', async (info) => {
    FetchMock.mockReturnValueOnce(
      new Promise<ElectronResult<VRCApiInfo>>((resolve) => {
        resolve({
          type: 'entity',
          cookies: [],
          result: { clientApiKey: 'testClientApiKey' },
        });
      }),
    );
    const customResult = render(<App />, {
      location: { pathname: info.path },
      route: { path: info.path },
      state: withAuthentication(),
    });
    await waitFor(() => {
      expect(FetchMock).toHaveBeenCalledWith({
        params: [{ key: 'apiKey', value: 'testApiKey' }],
        storedCookies: [{ key: 'auth', url: 'testUrl', value: 'testAuthToken' }],
        url: 'https://api.vrchat.cloud/api/1/config',
      });
    });

    await waitFor(() => {
      expect(UserOverviewMock).toHaveBeenCalled();
      expect(info.component).toHaveBeenCalled();
    });

    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
  });
});
