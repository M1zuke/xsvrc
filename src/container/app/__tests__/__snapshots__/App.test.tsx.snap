// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Component renders { component: [Function], path: '/' } Component 1`] = `
"<div class=\\"Component\\">
  <div class=\\"FloatingPath\\">/</div>
  <div class=\\"PageContent\\">
    <div data-testid=\\"home\\"></div>
  </div>
  <div data-testid=\\"user-overview\\"></div>
</div>"
`;

exports[`<App /> Component renders { component: [Function], path: '/friends' } Component 1`] = `
"<div class=\\"Component\\">
  <div class=\\"FloatingPath\\">/friends</div>
  <div class=\\"PageContent\\">
    <div data-testid=\\"friends\\"></div>
  </div>
  <div data-testid=\\"user-overview\\"></div>
</div>"
`;

exports[`<App /> Component renders login Component 1`] = `
"<div class=\\"Component\\">
  <div class=\\"FloatingPath\\">/login</div>
  <div data-testid=\\"login\\"></div>
</div>"
`;
